lib/java/spring/events/src/main/java/co/airy/spring/events/filters/HttpLogFilter.java

```diff
@@ -0,0 +1,76 @@
+package co.airy.spring.events.filters;
+
+import co.airy.spring.events.custom.HttpEventPublisher;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.server.ServletServerHttpRequest;
+import org.springframework.stereotype.Component;
+import org.springframework.web.filter.OncePerRequestFilter;
+import org.springframework.web.util.ContentCachingRequestWrapper;
+
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Component
+public class HttpLogFilter extends OncePerRequestFilter {
+
+    @Autowired
+    HttpEventPublisher httpEventPublisher;
+
+    private static final int MAX_JSON_LENGTH = 2048;
+
+    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
+        boolean isFirstRequest = !this.isAsyncDispatch(request);
+        HttpServletRequest requestToUse = request;
+        if (isFirstRequest && !(request instanceof ContentCachingRequestWrapper)) {
+            requestToUse = new ContentCachingRequestWrapper(request, MAX_JSON_LENGTH);
+        }
+
+        try {
+            filterChain.doFilter(requestToUse, response);
+        } finally {
+            publishEvent(requestToUse);
+        }
+    }
+
+    private void publishEvent(HttpServletRequest request) {
+        final Map<String, String> requestHeaders = getRequestHeaders((ContentCachingRequestWrapper) request);
+        final String requestUri = request.getRequestURL().toString();
+        String requestBody = null;
+
+        if ("POST".equalsIgnoreCase(request.getMethod())) {
+            requestBody = getRequestBody((ContentCachingRequestWrapper) request);
+        }
+        httpEventPublisher.publishCustomEvent(requestBody, requestHeaders, requestUri);
+    }
+
+    private Map<String, String> getRequestHeaders(ContentCachingRequestWrapper request) {
+        return (new ServletServerHttpRequest(request)).getHeaders().toSingleValueMap()
+                .entrySet()
+                .stream()
+                .map((entry) -> Map.entry(entry.getKey().toLowerCase(), entry.getValue()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    private String getRequestBody(ContentCachingRequestWrapper request) {
+        final byte[] buf = request.getContentAsByteArray();
+        final int bufferSize = buf.length;
+
+        if (bufferSize == 0) {
+            return null;
+        }
+
+        String payload;
+        try {
+            payload = new String(buf, 0, bufferSize, request.getCharacterEncoding());
+        } catch (Exception e) {
+            return null;
+        }
+
+        return payload;
+    }
+}
```

Comment:
I would have expected the filter to be in spring web. `lib/spring/events` sounds too generic to have an http log filter to me.

So I'm thinking we can:

- move the filter to web (and let web depend on events and nothing else which is probably less error prone than what we do now as you can't "forget" the dep)
- merge this lib with web
- keep the code as is but specialise the name of the lib to `lib/spring/httplog`

I think my favourite is the first option but two also works. I'm not so fond of option 3 unless we do not add this as a direct dep but package it in the `base_web_app` deps set